// arctic/dist/providers/apple.js
import { TimeSpan, createDate } from "oslo";
import { decodeBase64 } from "oslo/encoding";
import { createJWT } from "oslo/jwt";
import { OAuth2Client } from "oslo/oauth2";
const authorizeEndpoint = "https://appleid.apple.com/auth/authorize";
const tokenEndpoint = "https://appleid.apple.com/auth/token";
export class Apple {
    client;
    credentials;
    constructor(credentials, redirectURI) {
        this.client = new OAuth2Client(credentials.clientId, authorizeEndpoint, tokenEndpoint, {
            redirectURI
        });
        this.credentials = credentials;
    }
    async createAuthorizationURL(state, options) {
        return await this.client.createAuthorizationURL({
            state,
            scopes: options?.scopes
        });
    }
    async validateAuthorizationCode(code) {
        const result = await this.client.validateAuthorizationCode(code, {
            authenticateWith: "request_body",
            credentials: await this.createClientSecret()
        });
        const tokens = {
            accessToken: result.access_token,
            refreshToken: result.refresh_token ?? null,
            accessTokenExpiresAt: createDate(new TimeSpan(result.expires_in, "s")),
            idToken: result.id_token
        };
        return tokens;
    }
    async refreshAccessToken(refreshToken) {
        const result = await this.client.refreshAccessToken(refreshToken, {
            authenticateWith: "request_body",
            credentials: await this.createClientSecret()
        });
        const tokens = {
            accessToken: result.access_token,
            accessTokenExpiresAt: createDate(new TimeSpan(result.expires_in, "s")),
            idToken: result.id_token
        };
        return tokens;
    }
    async createClientSecret() {
        const audience = "https://appleid.apple.com";
        const payload = {};
        const jwt = await createJWT("ES256", parsePKCS8PEM(this.credentials.certificate), payload, {
            headers: {
                kid: this.credentials.keyId
            },
            issuer: this.credentials.teamId,
            includeIssuedTimestamp: true,
            expiresIn: new TimeSpan(5, "m"),
            audiences: [audience],
            subject: this.credentials.clientId
        });
        return jwt;
    }
}
function parsePKCS8PEM(pkcs8) {
    return decodeBase64(pkcs8
        .replace(/-----BEGIN PRIVATE KEY-----/, "")
        .replace(/-----END PRIVATE KEY-----/, "")
        .trim());
}
