// arctic/dist/providers/atlassian.js
import { OAuth2Client } from "oslo/oauth2";
import { TimeSpan, createDate } from "oslo";
const authorizeEndpoint = "https://auth.atlassian.com/authorize";
const tokenEndpoint = "https://auth.atlassian.com/oauth/token";
export class Atlassian {
    client;
    clientSecret;
    constructor(clientId, clientSecret, redirectURI) {
        this.client = new OAuth2Client(clientId, authorizeEndpoint, tokenEndpoint, {
            redirectURI
        });
        this.clientSecret = clientSecret;
    }
    async createAuthorizationURL(state, options) {
        const url = await this.client.createAuthorizationURL({
            state,
            scopes: options?.scopes
        });
        url.searchParams.set("audience", "api.atlassian.com");
        url.searchParams.set("prompt", "consent");
        return url;
    }
    async validateAuthorizationCode(code) {
        const result = await this.client.validateAuthorizationCode(code, {
            authenticateWith: "request_body",
            credentials: this.clientSecret
        });
        const tokens = {
            accessToken: result.access_token,
            accessTokenExpiresAt: createDate(new TimeSpan(result.expires_in, "s")),
            refreshToken: result.refresh_token ?? null
        };
        return tokens;
    }
    async refreshAccessToken(refreshToken) {
        const result = await this.client.refreshAccessToken(refreshToken, {
            authenticateWith: "request_body",
            credentials: this.clientSecret
        });
        const tokens = {
            accessToken: result.access_token,
            accessTokenExpiresAt: createDate(new TimeSpan(result.expires_in, "s")),
            refreshToken: result.refresh_token ?? null
        };
        return tokens;
    }
}
