// arctic/dist/providers/auth0.js
import { OAuth2Client } from "oslo/oauth2";
export class Auth0 {
    client;
    appDomain;
    clientSecret;
    constructor(appDomain, clientId, clientSecret, redirectURI) {
        this.appDomain = appDomain;
        const authorizeEndpoint = this.appDomain + "/authorize";
        const tokenEndpoint = this.appDomain + "/oauth/token";
        this.client = new OAuth2Client(clientId, authorizeEndpoint, tokenEndpoint, {
            redirectURI
        });
        this.clientSecret = clientSecret;
    }
    async createAuthorizationURL(state, options) {
        const scopes = options?.scopes ?? [];
        return await this.client.createAuthorizationURL({
            state,
            scopes: [...scopes, "openid"]
        });
    }
    async validateAuthorizationCode(code) {
        const result = await this.client.validateAuthorizationCode(code, {
            credentials: this.clientSecret
        });
        const tokens = {
            accessToken: result.access_token,
            refreshToken: result.refresh_token,
            idToken: result.id_token
        };
        return tokens;
    }
    async refreshAccessToken(refreshToken) {
        const result = await this.client.refreshAccessToken(refreshToken, {
            credentials: this.clientSecret
        });
        const tokens = {
            accessToken: result.access_token,
            refreshToken: result.refresh_token,
            idToken: result.id_token
        };
        return tokens;
    }
}
