// sveltekit-superforms/dist/adapters/arktype.js
import { type } from 'arktype';
import { createAdapter, createJsonSchema } from './adapters.js';
import { memoize } from '../memoize.js';
async function _validate(schema, data) {
    const result = schema(data);
    if (!(result instanceof type.errors)) {
        return {
            data: result,
            success: true
        };
    }
    const issues = [];
    for (const error of result) {
        issues.push({ message: error.message, path: error.path });
    }
    return {
        issues,
        success: false
    };
}
function _arktype(schema, options) {
    return createAdapter({
        superFormValidationLibrary: 'arktype',
        defaults: options.defaults,
        jsonSchema: createJsonSchema(options),
        validate: async (data) => _validate(schema, data)
    });
}
function _arktypeClient(schema) {
    return {
        superFormValidationLibrary: 'arktype',
        validate: async (data) => _validate(schema, data)
    };
}
export const arktype = /* @__PURE__ */ memoize(_arktype);
export const arktypeClient = /* @__PURE__ */ memoize(_arktypeClient);
